sources/Algo/src/Algo.cpp

```diff
@@ -74,41 +74,44 @@ MainConnexComponentAlgorithm::operator()(const NodePtr& node) {
-    serviceManager_(serviceManager) {}
+    serviceManager_{serviceManager},
+    buses_{buses} {}
 
 void
 GeneratorDefinitionAlgorithm::operator()(const NodePtr& node) {
   auto& node_generators = node->generators;
-  if (node_generators.size() == 1) {
-    auto model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN;
-    if (IsOtherGeneratorConnectedBySwitches(node)) {
-      // Use the model with impedance in case another generator is connected to the node through a switch network path
-      model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN
-                                         : GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN;
-    }
-
-    const auto& gen = node_generators.front();
-
-    if ((model == GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN || model == GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN) &&
-        !isDiagramValid(gen)) {
-      return;
-    }
-
-    generators_.emplace_back(gen.id, model, node->id, gen.points, gen.qmin, gen.qmax, gen.pmin, gen.pmax);
-  } else {
-    auto model =
-        useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::WITH_IMPEDANCE_SIGNALN : GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN;
-    for (auto it = node_generators.begin(); it != node_generators.end(); ++it) {
-      if (model == GeneratorDefinition::ModelType::WITH_IMPEDANCE_DIAGRAM_PQ_SIGNALN && !isDiagramValid(*it)) {
-        continue;
+  for (const auto& generator : node_generators) {
+    auto it = busMap_.find(generator.regulatingBusId);
+    auto nbOfRegulatingGenerators = it->second;
+    GeneratorDefinition::ModelType model;
+    if (node_generators.size() == 1 && IsOtherGeneratorConnectedBySwitches(node)) {
+      model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::REMOTE_SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_REMOTE_SIGNALN;
+    } else if (nbOfRegulatingGenerators == dfl::inputs::NetworkManager::nbOfRegulatingGenerators::ONE) {
+      if (generator.regulatingBusId == generator.connectedBusId) {
+        model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_SIGNALN;
+      } else {
+        model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::REMOTE_SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_REMOTE_SIGNALN;
       }
-      generators_.emplace_back(it->id, model, node->id, it->points, it->qmin, it->qmax, it->pmin, it->pmax);
+    } else if (nbOfRegulatingGenerators == dfl::inputs::NetworkManager::nbOfRegulatingGenerators::MULTIPLES) {
+      model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::PROP_SIGNALN : GeneratorDefinition::ModelType::DIAGRAM_PQ_PROP_SIGNALN;
+      buses_.insert({generator.regulatingBusId, generator.id});
+    } else {
+      //Not supposed to be here
```

Comment:
Use a switch to avoid this else case => show better that it is an enum, which the name doesn't